types/react-keyboard-event-handler/KeyboardEventHandler.d.ts

```diff
@@ -0,0 +1,14 @@
+import * as React from "react";
+
+export interface KeyboardEventHandlerProps {
+    handleKeys?: readonly string[];
+    handleEventType?: "keydown" | "keyup" | "keypress";
+    handleFocusableElements?: boolean;
+    onKeyEvent?: (key: string, event: KeyboardEvent) => void;
+    isDisabled?: boolean;
+    isExclusive?: boolean;
+    children?: React.ReactNode;
+}
+
+declare const KeyboardEventHandler: React.SFC<KeyboardEventHandlerProps>;
+export default KeyboardEventHandler;
```

Comment:
1. There’s no `KeyboardEventHandler.js` file in react-keyboard-event-handler, so this file shouldn’t exist. The types go in `index.d.ts`.
2. react-keyboard-event-handler is a UMD module, not an ES module with a default export. The correct form is

```ts
import * as React from "react";

namespace KebyboardEventHandler {
  export interface KeyboardEventHandlerProps {
      handleKeys?: readonly string[];
      handleEventType?: "keydown" | "keyup" | "keypress";
      handleFocusableElements?: boolean;
      onKeyEvent?: (key: string, event: KeyboardEvent) => void;
      isDisabled?: boolean;
      isExclusive?: boolean;
      children?: React.ReactNode;
  }
}

declare const KeyboardEventHandler: React.SFC< KeyboardEventHandler.KeyboardEventHandlerProps>;
export = KeyboardEventHandler;
export as namespace KeyboardEventHandler;
```

Reply:
RE 2. Sorry, I'm not entirely sure I follow. If I define it as a default export then my personal build works and I can import it as intended:

```typescript
import KeyboardEventHandler from "react-keyboard-event-handler";
```

If I define it as you said, then it completely breaks unless I turn on `esModuleInterop` or use a `require` style import.

Reply:
Okay, I think I've got it now, I needed to use this pattern:

```typescript
type KeyboardEventHandler = React.SFC<KeyboardEventHandler.KeyboardEventHandlerProps>;

declare const KeyboardEventHandler: KeyboardEventHandler & {
    default: KeyboardEventHandler;
};

export = KeyboardEventHandler;
```

Reply:
> If I define it as you said, then it completely breaks unless I turn on `esModuleInterop` or use a `require` style import.

This is correct and expected behavior. You should enable `esModuleInterop` if you intend to import this library through a default import. You should be using a `require` import here on DefinitelyTyped. Your update is incorrect because there is no `default` property on the export: https://runkit.com/andrewbranch/60673fbc60b076001bc0c7f4

Reply:
The react-keyboard-event-handler package's README examples use default import styles.

Also, it's source index.js file is written as:

```javascript
export { default } from "./KeyboardEventHandler";
```

The types I have written, in the latest version, do work in Typescript and then at runtime, for both require and default style imports, with or without esModuleInterop set.

The types you are asking for work worse for me, requiring the wrong import type, or a compiler flag to be set.

The latest types code simply takes from a pattern I noticed in multiple other types packages.

So, I really do not understand why what you are suggesting is better. Please, if you could, point me to some documentation/articles, so that I can learn why and the quirks in play here.

Reply:
Your job on DefinitelyTyped is to describe the shape of the module found at the `main` field of react-keyboard-event-handler’s package.json, which is [this file](https://unpkg.com/browse/react-keyboard-event-handler@1.5.4/lib/react-keyboard-event-handler.js). What you find in the `src` directory is completely irrelevant, as is the import style used by the library’s README or in your own projects.

> The types I have written, in the latest version, do work in Typescript and then at runtime, for both require and default style imports, with or without esModuleInterop set.

Your latest version does put the correct thing at both the CJS export and the `default` named export. This is _pretty_ good—much better than the first attempt. But you have achieved this at the cost of implying that the package, when `require`’d, has a property called `default`, which it does not:

<img width="381" alt="image" src="https://user-images.githubusercontent.com/3277153/113448082-073e9080-93b0-11eb-91e6-0150c626fa5e.png">

<img width="740" alt="image" src="https://user-images.githubusercontent.com/3277153/113448243-5258a380-93b0-11eb-8509-21c5e16bb177.png">

The problem is that these typings do not correctly catch this as an error.

> The types you are asking for work worse for me, requiring the wrong import type, or a compiler flag to be set.

A default import is the _wrong import type_ if you don’t set `--esModuleInterop`. If you don’t set `--esModuleInterop`, you are saying “Please do not let me import CJS modules (like react-keyboard-event-handler) with a default import, because my runtime does not support that kind of interoperability, and I do not want TypeScript’s help attaining that kind of interoperability. I want my imports to exactly match my exports, so give me errors when I’ve written a mismatch.” This is an uncommon thing to want—most users should enable `--esModuleInterop`. But when you are writing definitions on DT, you must leave this decision up to the users.

> The latest types code simply takes from a pattern I noticed in multiple other types packages.

Some libraries package their code with export interop which _does_ in fact set a `default` property on the CJS export. react-keyboard-event-handler is not one of those libraries, but if it were, what you have now would be correct. It is also very likely that many packages on DT are incorrect.

> Please, if you could, point me to some documentation/articles, so that I can learn why and the quirks in play here.

There is no explanation of this out there that I am happy with. I need to write one. The relevant authority is [here](https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html), but I don’t particularly like how it’s represented. It implies that it’s generally ok for `module.exports =` in JS can to be represented by `export default` in a declaration file, but that you shouldn’t do this on DT. I would personally say that it’s totally wrong, and you should never do it anywhere, especially on DT, and it’s a coincidence that it will sometimes work for some users.

Reply:
Okay. I will change it then if that is considered best practice and safest. But, I must say that the knowledge and advice available online on this subject is a contradictory, badly documented mess. And I'm still unhappy with this being the first package I've ever used, in several years of full time TS development, that will have typings that force me to enable that compiler flag or switch to an old fashioned `require` style import. 

And can you tell me, what changes would somebody (ie. me) need to make to the original package, so it no longer has this problem?

Reply:
The original package doesn’t have a problem. It’s using the world’s most compatible form of export. If it were to switch to using a proper ESM default export, you would be correct to use a default import without needing a compiler flag. But anyone using `require` in that case would have to switch to doing `require('react-keyboard-event-handler').default`.

If you’ve been developing with React and TypeScript for a few years, I’d be surprised if you’ve never encountered this before. Ever notice how TypeScript users say `import * as React from 'react'` while everyone else says `import React from 'react'`? That’s the exact same issue. Without `esModuleInterop` enabled, `import *` can be used instead of `import ... = require(...)`. And if you _haven’t_ ever seen `import * as React` then you’re already using `esModuleInterop` or `allowSyntheticDefaultImports`.

> I'm still unhappy

I’m unhappy that you’re unhappy, because I think this is a good thing! `esModuleInterop` is hands-down one of the best compiler flags there is. Almost everyone should use it. It’s baked into `tsc --init`. If we had known more about how the ES module spec would evolve when we first started supporting ES imports, it might have been the default behavior in TypeScript. There will probably probably a future module resolution mode where it _is_ the default, to reflect the interop built into Node 14 (and modern browsers?).

> I must say that the knowledge and advice available online on this subject is a contradictory, badly documented mess.

I know. I’m now fully inspired to write an authoritative blog post about this in the coming weeks.

Happy Friday!

Reply:
That is exactly my issue though. I would normally use `import * as Foo from "foo"` in these cases, but if I do that with this package, then it fails at runtime.

Reply:
It fails at runtime with `import *` only if you enable `esModuleInterop`. If you enable `esModuleInterop`, you must use a default import for this library. If you leave `esModuleInterop`/`allowSyntheticDefaultImports` off, you can use `import ... = require('...')` or `import *`.

Reply:
Thank you sir for all your help. I think I have a good understanding of this now, that will help in future. And, additionally, I have got to the root of my confusion: at work we use Webpack, but in my personal testing of this module I was using Parcel, which - as it turns out - implicitly applies the `esModuleInterop` flag, whereas the type-checker was still actually looking at the value in my tsconfig. Thank you for your great insight.